Open API 로직 변경 및 예외 처리하기

koreandroid opened the pull request:
- getAbandonmentPublic 함수에 pageNo 쿼리 매개변수를 추가하였습니다.
<br>&nbsp;
- numOfRows 쿼리 매개변수의 기본값을 40으로 설정하였습니다.
<br>&nbsp;
- 입양 목록 화면에 보여줄 아이템들을 AdoptRecyclerViewAdapter 클래스의 dataSet 프로퍼티에서 관리하도록 하였습니다.
<br>&nbsp;
- 어댑터가 목록의 마지막 아이템을 보여줄 때, pageNo 프로퍼티의 값을 1만큼 증가시킨 후 새로운 API 요청을 수행하도록 로직을 변경하였습니다.
<br>&nbsp;
- java.net.SocketTimeoutException을 catch 블록으로 처리하였습니다.
<br>&nbsp;
- 접수일을 비교하여 목록의 날짜가 역전되는 현상을 방지하도록 했습니다.

koreandroid committed:
rename: AdoptAdapter.kt 파일 AdoptRecyclerViewAdapter.kt로 이름 변경

koreandroid committed:
feat: 예외 및 오류 상황에서 null이 되는 변수 타입 변경

koreandroid committed:
feat: 입양 목록 화면 Open API 로직 변경

- getAbandonmentPublic 함수에 pageNo 쿼리 매개변수를 추가함.
- numOfRows 쿼리 매개변수의 기본값을 40으로 설정함.
- 입양 목록 화면에 보여줄 아이템들을 AdoptRecyclerViewAdapter 클래스의 dataSet 프로퍼티에서 관리하도록 함.
- 어댑터가 목록의 마지막 아이템을 보여줄 때, pageNo 프로퍼티의 값을 1만큼 증가시킨 후 새로운 API 요청을 수행하도록 로직을 변경함.
- java.net.SocketTimeoutException을 catch 블록으로 처리함.

koreandroid committed:
fix: 예외 처리 시의 pageNo 프로퍼티 값 변경 오류 수정

koreandroid committed:
feat: 접수일을 비교하는 로직 추가

- 조건문을 추가하여 dataSet이 변경되었을 경우에만 notifyItemRangeInserted 함수를 호출하도록 함.
- 접수일을 비교해 abandonmentInfos를 정리하는 확장함수를 추가함.

koreandroid committed:
refactor: 불필요한 스코프 함수 제거

- 확장함수를 호출하여 사용하고 있는 사실이 명확하게 드러나도록 with 함수를 제거함.
- 불필요하게 활용하고 있던 run 함수를 제거함.

Ju-lotty commented:
재성님 기능개선 하시느라 고생 많으셨습니다! 점심시간 이후 바로 코맨트 남기겠습니다!

Ju-lotty commented on app/src/main/java/com/teamtriad/forpets/data/source/network/AdoptService.kt:
```diff
@@ -18,7 +18,8 @@ interface AdoptService {
         @Query("bgnde") bgnde: String,
         @Query("endde") endde: String,
-        @Query("numOfRows") numOfRows: String
+        @Query("pageNo") pageNo: String,
+        @Query("numOfRows") numOfRows: String = "40",
```
너무 많은 데이터를 가져오는 것보다 데이터를 가져오는 시간이 줄겠네요~!!

Ju-lotty commented on app/src/main/java/com/teamtriad/forpets/ui/adopt/adapter/AdoptRecyclerViewAdapter.kt:
```diff
@@ -22,10 +36,67 @@ class AdoptAdapter(private val dataSet: List<AbandonmentInfo>) :
+                            ?.item
+                    ).let {
+                        notifyItemRangeInserted(dataSet.size - it, it)
+                    }
```
if문과 let을 분리하신 이유가 있을까요? Resopnse 상태가 200이라면 괜찮지 않을까요?

koreandroid replied:
Outdated된 코드이기 때문에 [32408fe](https://github.com/sesac-team-triad/for-pets-android/pull/29/commits/32408fe81f368775b284a0ba1f36a4e403c1842f) 커밋을 기준으로 말씀드리겠습니다.
먼저 updataDataSet 함수에 null 또는 빈 리스트가 전달될 수 있는 상황이 있기 때문에, undataDataSet 함수가 새롭게 추가된 정보의 개수를 반환하도록 하였습니다. 만약에 새롭게 추가된 정보가 없다면 notifyItemRangedInserted 함수를 호출할 필요가 없기 때문에 if 문을 추가하였고, 반환값을 가지고 다음으로 if 문이 실행되는 점을 보다 명확히 나타내기 위해 run 함수를 사용하였습니다.
유기동물 조회 서비스 API의 경우, Retrofit을 활용한 네트워크 통신이 성공적이었다(response.isSuccessful) 하더라도 요청 메시지에 따라 오류가 발생할 경우에는 body(HTTP GET 요청에 대한 응답의 body가 아니라) 없이 header(마찬가지로 HTTP 응답의 헤더가 아니라)만 포함된 응답 메시지를 성공적인 HTTP 응답으로 보내주는 것 같습니다. 따라서 JSON 형식으로 되어 있는 HTTP 응답의 본문 response.body()!!가 AbandonmentPublic.response.body 데이터 클래스로 파싱될 때 null이 될 수 있고, 보내줄 데이터가 없는 경우에는 items 객체 안의 item 필드가 아예 생략되기 때문에, 마찬가지로 null이 될 수 있습니다.
정리하자면, HTTP GET을 통한 네트워크 요청은 성공적이더라도 위와 같은 상황들에서는 decode된 결과가 null인 경우들이 있을 수 있고, 아이템 리스트가 있는 상황에서도 updateDataSet 함수를 통해 추가된 아이템이 0개일 수 있기 때문에 상황에 맞게 notifyItemRangedInserted 함수를 호출할 수 있도록 구현하였습니다.

Ju-lotty commented on app/src/main/java/com/teamtriad/forpets/ui/adopt/adapter/AdoptRecyclerViewAdapter.kt:
```diff
@@ -99,6 +98,26 @@ class AdoptRecyclerViewAdapter(private val lifecycleScope: LifecycleCoroutineSco
+        if (dataSet.isEmpty()) return
+
+        while (isNotEmpty()) {
+            if (dataSet[dataSet.lastIndex].happenDate >= this[0].happenDate) break
+
+            removeFirst()
+        }
+    }
+
+    private fun MutableList<AbandonmentInfo>.adjustItemDuplicate() {
+        while (isNotEmpty()) {
+            if (dataSet.find {
+                    it.noticeNo == this[0].noticeNo
+                } == null) break
+
+            removeFirst()
+        }
+    }
```
updateDataSet 함수에서 쓰일 adjustItemDuplicate과 adjustItemInversion를 합치기보다 목적을 나누어 보니 보기편하네요~~

Ju-lotty reviewed (commented):
코맨트 남깁니다!

Ju-lotty commented on app/src/main/java/com/teamtriad/forpets/ui/adopt/adapter/AdoptRecyclerViewAdapter.kt:
```diff
@@ -22,10 +36,67 @@ class AdoptAdapter(private val dataSet: List<AbandonmentInfo>) :
+        }
+    }
+
+    private fun updateDataSet(item: List<AbandonmentInfo>?): Int {
+        if (item.isNullOrEmpty()) return 0
```
아이템이 없거나 Null일 경우에 대한 UI처리도 필요해보입니다!

koreandroid replied:
[73aa7c7](https://github.com/sesac-team-triad/for-pets-android/pull/29/commits/73aa7c7127903c56c83f2404da680c82d6331225) 커밋을 작성하였습니다. 감사합니다.

koreandroid committed:
feat: 새로운 API 요청 수행 시점 및 numOfRows 기본값 변경

API 요청이 자연스럽게 이루어지도록 새로운 요청을 수행하는 시점을 변경하고 numOfRows 쿼리 매개변수의 기본값을 40에서 30으로 변경함.

koreandroid committed:
feat: 타임아웃 시간 변경

koreandroid committed:
add: 스트링 리소스 추가

koreandroid committed:
feedback: Open API 요청 후 목록이 생성되지 않았을 때 처리 및 tv_fallback 뷰 UI 추가

koreandroid committed:
fix: adjustItemDuplicate() 오류 수정

koreandroid committed:
design: 입양 목록 아이템 내용 여백 변경

koreandroid committed:
feat: 타임아웃 시간 변경
